include/picongpu/particles/collision/binary/DefaultAlg.hpp

```diff
@@ -0,0 +1,499 @@
+                * comsMomentumMag0;
+            val = val / ( coeff0 * coeff1 * gammaComs );
+            return val;
+
+        }
+        // Stared gamma times mass.
+        static DINLINE float_X coeff(
+            float3_X labMomentum,
+            float_X mass,
+            float_X gamma,
+            float_X gammaComs,
+            float3_X comsVelocity
+        )
+        {
+            float_X dot = pmacc::math::dot( comsVelocity, labMomentum );
+            float_X val = mass * gamma - dot / ( c*c );
+            return gammaComs * val;
+        }
+
+        static DINLINE float_X calcCosXi(
+            float_X const s12,
+            float_X const u
+        )
+        {
+            // TODO: find some better way to restrict the val to [-1, +1].
+            if ( s12 < 0.1_X )
+            {
+                float_X cosXi = 1.0_X + s12 * math::log( u );
+                if ( cosXi < -1.0_X || cosXi > 1.0_X )
+                {
+                    if  (cosXi < -1.0_X ) return -1.0_X;
+                    else  return 1.0_X;
+                }
+                else return cosXi;
+            }
+            else if ( s12 < 6.0_X )
+            {
+                float_X a;
+                if( s12 < 3.0_X )
+                {
+                    float_X s12sq = s12 * s12;
+                    a = 0.0056958_X + 0.9560202_X * s12 - 0.508139_X * s12sq +
+                        0.47913906_X * s12sq * s12 - 0.12788975_X * s12sq *
+                        s12sq + 0.02389567_X * s12sq * s12sq * s12sq;
+                    a = 1/a;
+                }
+                else
+                {
+                    a = 3.0_X * math::exp( -1.0_X * s12 );
+                }
+                float_X bracket = math::exp( -1.0_X * a) +
+                    2.0_X * u * std::sinh( a );
+                float_X cosXi = math::log( bracket ) / a;
+                if ( cosXi < -1.0_X || cosXi > 1.0_X )
+                {
+                    if  (cosXi < -1.0_X ) return -1.0_X;
+                    else return 1.0_X;
+                }
+                else return cosXi;
+            }
+            else
+            {
+                // + 1 in Perez 2012 but cos in [-1,1]. Smilei uses -1.
+                return 2 * u - 1;
+            }
+        }
+
+        // (12) in [Perez 2012]
+        static DINLINE float3_X calcFinalComsMomentum(
+            float3_X const p,
+            float_X const cosXi,
+            float_X const phi
+        )
+        {
+            float_X sinPhi, cosPhi;
+            pmacc::math::sincos( phi, sinPhi, cosPhi);
+            float_X sinXi = math::sqrt( 1 - cosXi * cosXi );
+
+            float3_X finalVec;
+            float_X const pAbs = math::sqrt( pmacc::math::abs2( p ) );
+            float_X const pPerp = math::sqrt( p.x( ) * p.x( ) + p.y( )
+                * p.y( ) );
+            // TODO chose a better limmit?
+            if( pPerp > std::max( std::numeric_limits< float_X >::epsilon, 1.0e-10_X ) *  pAbs )
```

Comment:
if `pAbs` == 0.0 than `pPerp` < epsilon will be allowed too and can result in `NaN` later.

Reply:
I reversed the the condition and the content of if and else. So for `pAbs = pPerp = 0` the result is (0,0,0).  An alternative would be to leave the momentum unchanged in that case.